fix(engine): open single message or signal subscription for the same event

skayliu opened the pull request:
## Description

When a user deploys the same process multiple times without changes, no additional message or signal start event subscriptions will be opened, if there are top-level message or signal start events present in the process.

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #11308 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

korthout commented on engine/src/test/java/io/camunda/zeebe/engine/processing/message/MessageStartEventSubscriptionTest.java:
```diff
@@ -280,13 +282,35 @@ public void shouldHaveSameSubscriptionKey() {
+
+    engine.deployment().withXmlResource(process).deploy();
+
+    // when
+    engine.deployment().withXmlResource(process).deploy();
+
+    // then
+    assertThat(RecordingExporter.messageStartEventSubscriptionRecords(
+            MessageStartEventSubscriptionIntent.CREATED)
+        .withBpmnProcessId("processId")
+        .limit(2))
+        .extracting(Record::getValue)
+        .extracting(MessageStartEventSubscriptionRecordValue::getMessageName,
+            MessageStartEventSubscriptionRecordValue::getStartEventId)
+        .hasSize(1)
+        .containsExactly(tuple(MESSAGE_NAME1, EVENT_ID1));
```
❌ Since the fix, this `limit(2)` will never hit the second record because only one such record is recorded. This means the test will be very slow because it waits for the record stream to time out (5 seconds).

🤔 A fundamental question pops up when attempting to resolve this problem. How can we limit the recorded stream to a record that does not exist on the stream? I.e., the second subscription record does not exist on the stream, so how else can we limit the log? 

💡 There is no good answer, but there is a trick we could try. By adding another command that will be rejected (after the command that we want to test), we can use the position of the command to limit the stream. That means we now have all the records on the stream from the start to the end, plus the rejected command we added additionally. Of course, this makes the test worse. However, sometimes that may be a cost we can pay when it helps us avoid a bug. That would look something like this:

```java
// when
engine.deployment().withXmlResource(process).deploy();

// then
// deploy an empty deployment, so we can use the position to limit the recorded stream
final var position = engine.deployment().expectRejection().deploy().getPosition();
assertThat(
        RecordingExporter.records()
            .limit(r -> r.getPosition() >= position)
            .filter(r -> r.getValueType() == ValueType.MESSAGE_START_EVENT_SUBSCRIPTION))
    .describedAs("Expect only one message start event subscription for duplicate deployments")
    .hasSize(1);
```

korthout reviewed (changes requested):
Thanks for this fix @skayliu 🚀 

👍 Nice and simple solution

👍 Your test cases cover the situation where the bug happens and failed when the bug was still present

❌ The applied limit no longer works after the fix was applied, slowing down the tests. I've tried to describe a solution that is not pretty but helps to speed up the test again.

Please have a look

skayliu commented:
Hey, @korthout , I've applied you hints, please reivew it again, thank you.

skayliu commented:
@korthout, There is a spotless error again, I've run `mvn spotless:apply` again. A couple of days ago i got my IntelliJ IDEA updated to v2022.3.3, Sorry about that.

korthout commented:
>There is a spotless error again, I've run mvn spotless:apply again. A couple of days ago i got my IntelliJ IDEA updated to v2022.3.3, Sorry about that.

@remcowesterhoud I believe you also ran into issues with that version right? Did you find a solution to this?

remcowesterhoud commented:
> > There is a spotless error again, I've run mvn spotless:apply again. A couple of days ago i got my IntelliJ IDEA updated to v2022.3.3, Sorry about that.
> 
> @remcowesterhoud I believe you also ran into issues with that version right? Did you find a solution to this?

For me it was when I switched from IntelliJ CE to a licensed version. Anyway, what happened was that the google-java-format plugin was broken. I had to add some jvm options to resolve it and this worked for my on Mac.

1. IntelliJ -> Help -> Edit Custom VM Options
2. Add:
```
--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED
--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED
--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED
--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED
--add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED
--add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED
--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED
--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED
--add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED
--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED
```

skayliu commented:
> > > There is a spotless error again, I've run mvn spotless:apply again. A couple of days ago i got my IntelliJ IDEA updated to v2022.3.3, Sorry about that.
> > 
> > 
> > @remcowesterhoud I believe you also ran into issues with that version right? Did you find a solution to this?
> 
> For me it was when I switched from IntelliJ CE to a licensed version. Anyway, what happened was that the google-java-format plugin was broken. I had to add some jvm options to resolve it and this worked for me on Mac.
> 
> 1. IntelliJ -> Help -> Edit Custom VM Options
> 2. Add:
> 
> ```
> --add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED
> --add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED
> --add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED
> ```



> > > There is a spotless error again, I've run mvn spotless:apply again. A couple of days ago i got my IntelliJ IDEA updated to v2022.3.3, Sorry about that.
> > 
> > 
> > @remcowesterhoud I believe you also ran into issues with that version right? Did you find a solution to this?
> 
> For me it was when I switched from IntelliJ CE to a licensed version. Anyway, what happened was that the google-java-format plugin was broken. I had to add some jvm options to resolve it and this worked for me on Mac.
> 
> 1. IntelliJ -> Help -> Edit Custom VM Options
> 2. Add:
> 
> ```
> --add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED
> --add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED
> --add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED
> --add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED
> ```

These are mine, sounds like i missed some args.
```
--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED
--add-exports=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED
--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED
--add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED
--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED
--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED
```

remcowesterhoud commented:
@skayliu Potentially you don't need all of them. I just copied them from somewhere 😄 But you can try if it works.

skayliu commented:
> @skayliu Potentially you don't need all of them. I just copied them from somewhere 😄 But you can try if it works.

I just add them all, may it works 🙏 .

skayliu commented:
> @skayliu Potentially you don't need all of them. I just copied them from somewhere 😄 But you can try if it works.
> 
> An easy way to test this is to make a long line of code and save the file. If it reformats it to different lines you should be good. Also you probably have to restart IntelliJ after changing them.

@remcowesterhoud , Great, It worked,  Thanks a lot ☕ .

korthout reviewed (commented):
Great work @skayliu 👏 Thanks again

LGTM 👍 

bors merge

korthout reviewed (approved):
Oops, forgot to approve

bors merge
